rdf-delta-server-local/src/main/java/org/seaborne/delta/zk/direct/ValidZooKeeperSupplier.java

```diff
@@ -0,0 +1,253 @@
+     * @param connectString A comma-separated list of host:port pairs pointing to ZooKeeper Ensemble members.
+     * @param retries The number of times to try to connect to a ZooKeeper Ensemble before giving up.
+     * @throws IOException if a problem occurs while attempting to connect to the ZooKeeper Ensemble.
+     * @throws KeeperException if a problem occurs getting the current ZooKeeper Ensemble configuration.
+     * @throws InterruptedException if the current thread is interrupted while waiting.
+     */
+    public ValidZooKeeperSupplier(final CharSequence connectString, final int retries) throws IOException, KeeperException, InterruptedException {
+        this.connectString = connectString;
+        this.retries = retries;
+        this.connect();
+    }
+
+    /**
+     * Instantiates a new {@link ZooKeeper} instance and updates the Ensemble configuration.
+     * @throws IOException if a problem occurs while attempting to connect to the ZooKeeper Ensemble.
+     * @throws KeeperException if a problem occurs getting the current ZooKeeper Ensemble configuration.
+     * @throws InterruptedException if the current thread is interrupted while waiting.
+     */
+    private void connect() throws IOException, KeeperException, InterruptedException {
+        this.zooKeeper = new ZooKeeper(
+            this.connectString.toString(),
+            10_000,
+            watchedEvent -> {
+                synchronized (this.token) {
+                    switch (watchedEvent.getState()) {
+                        case SyncConnected:
+                        case SaslAuthenticated:
+                        case ConnectedReadOnly:
+                            this.isValid = true;
+                            break;
+                        case Closed:
+                        case Expired:
+                        case AuthFailed:
+                        case Disconnected:
+                            this.isValid = false;
+                    }
+                    this.token.notifyAll();
+                }
+            }
+        );
+        this.updateConfig();
+    }
+
+    @Override
+    public ZooKeeper get() {
+        try {
+            long tries = 1;
+            synchronized (this.token) {
+                while (!this.isValid) {
+                    if (this.isClosed) {
+                        throw new ZkException("ValidZooKeeperSupplier instance has been closed.");
+                    }
+                    // Unfortunately, the ZooKeeper states are not well-documented. This represents the
+                    // best guess with an incomplete understanding.
+                    LOG.info("Connection flagged as invalid with state: {}", this.zooKeeper.getState());
+                    switch (this.zooKeeper.getState()) {
+                        case CONNECTING:
+                            LOG.info("Waiting until connected...");
+                            this.token.wait(3000);
```

Comment:
If the node is healthy, connections are nearly instantaneous. If the node is unhealthy, it needs to get out of the way of the healthy nodes. I prefer Fail Fast for this since much of the robustness is provided by the load balancer and auto-scaling group.

Reply:
Is my answer above satisfactory?

Reply:
Not entirely, but it's not a priority.